### DISCUSSION TITLE
uiFixedSlider and custom values

### DISCUSSION THREAD

**Discussion by damiensellier** [RomanKubiak/ctrlr#255] [Q&A] [answered]
Hi Ctrlr people,

I am experimenting with the uiFixedSlider properties which is very promissing and will help me not making hundreds of tables.

For my current yamaha spx990 project I need to display values that reflect the algo paramaters of the device (levels in dB, frequencies, on/off...) instead of decimal values like 0,1,2 etc

Let's take an exemple:
For the compressor algo, the thresholds goes from -24dB to 12dB, with values from 
232 for -24dB, ... 
255 for -1dB, 
0 for 0dB, 
1 for +1dB .... 
12 for +12dB

I converted my uiSlider to uiFixedSlider and added the following values in the "slider content" field.

```
-24 dB=232
-23 dB=233
-22 dB=234
-21 dB=235
-20 dB=236
-19 dB=237
-18 dB=238
-17 dB=239
-16 dB=240
-15 dB=241
-14 dB=242
-13 dB=243
-12 dB=244
-11 dB=245
-10dB=246
-9 dB=247
-8 dB=248
-7 dB=249
-6 dB=250
-5 dB=251
-4 dB=252
-3 dB=253
-2 dB=254
-1 dB=255
0 dB=0
1 dB=1
2 dB=2
3 dB=3
4 dB=4
5 dB=5
6 dB=6
7 dB=7
8 dB=8
9 dB=9
10 dB=10
11 dB=11
12 dB=12
```

The value displayed doesn't reflects the byte value received. for exemple a byte value of 6 displays : "-18 dB" (7th position in the list starting from 0) not "6 dB" as I would like.

It's like my syntax is not right and doesn't take in account "=Value"  in the form "relatedText=value"

Am I missing something about this feature?

thanks a lot for your help

Damien

**Comment by dnaldoog** [comment, marked as answer]
Hi Damien,

I think you would have to use _lua_ tables with the MIDI in message as far as I can tell unless there's some clever way of changing those values back e.g. say 255 back into the value 23?

The code below would convert a message <code>f0 41 00 00 00 00 01 7f f7</code> into __23__.

<pre>
<code>
myMidiReceived = function(--[[ CtrlrMidiMessage --]] midi)
local msb=midi:getData():getByte(5)*128
local lsb=midi:getData():getByte(6)
local v=msb+lsb
local t={
["comp_232"]=0,
["comp_233"]=1,
["comp_234"]=2,
["comp_235"]=3,
["comp_236"]=4,
["comp_237"]=5,
["comp_238"]=6,
["comp_239"]=7,
["comp_240"]=8,
["comp_241"]=9,
["comp_242"]=10,
["comp_243"]=11,
["comp_244"]=12,
["comp_245"]=13,
["comp_246"]=14,
["comp_247"]=15,
["comp_248"]=16,
["comp_249"]=17,
["comp_250"]=18,
["comp_251"]=19,
["comp_252"]=20,
["comp_253"]=21,
["comp_254"]=22,
["comp_255"]=23,
["comp_0"]=24,
["comp_1"]=25,
["comp_2"]=26,
["comp_3"]=27,
["comp_4"]=28,
["comp_5"]=29,
["comp_6"]=30,
["comp_7"]=31,
["comp_8"]=32,
["comp_9"]=33,
["comp_10"]=34,
["comp_11"]=35,
["comp_12"]=36
}
panel:getModulatorByName("compressor"):setModulatorValue(t["comp_"..v],false,false,false)
end
</pre>
</code>

[Compressor Algo_1_0_Hell-O-Kitty_2021-04-14_19-31.zip](https://github.com/RomanKubiak/ctrlr/files/6310496/Compressor.Algo_1_0_Hell-O-Kitty_2021-04-14_19-31.zip)

**Comment by dobo365** [comment]
Hi Damien,
  maybe that you should use uiCombo instead and this will work.
I suppose you understood the difference between the value and the Midi value.
With a combo content as:
CHORUS=0
ENSEMBLE=38
PHASER=64
FLANGER=84
USER=127

Phaser is still modulator value 2 but will send Midi value 64 in the CC message associated to this combo (for example).
it will also react on Midi value 64 on CC receive and not on 2.
Modulator value 2 is in this case only internal.

**Reply by dnaldoog** [reply]
@dobo365  didn’t know that - thanks.

**Reply by dobo365** [reply]
We all learn from each other all the time and this is what I like in this Ctrlr stuff ;-)

**Reply by damiensellier** [reply]
Thanks guys for your replies. 
I think I'll need to work with tables anyway.
**setMappedValue** would have been the way to go but the function doesn't exist, that's a bummer.

The way my pannel works is limited to the spx990 available messages.
There's just 2 types of useful messages : 
1. sysex memory dump (with values nibbleized like 0v 0V .... this is why I get values over 127 because 0F 0F gives FF
2. cc controls for only 2 paramaters. the parameter number is stored in the sysex memory message on byte 135 and 141

There's no button emulation for "store preset" nor a dedicated sysex request to save an edited preset to a slot. 
So to be able to communicate I have have to : 
1. request a dump of the current used preset.
2. the sysex is decoded and all parameter values are mapped to the related modulators.
3. When the user click on a slider a sysex message is sent back to the device with the parameter number changed on byte 135 or 141 (it's just a 300bytes message)
4. the modulator changes the value, the value sent is a cc from 0 to 127 whatever the sysex values were.
5. the new parameter value replace the older in the memoryblock (not 0 to 127 but the actual value)
6. if the user click another parameter the new memoryblock is re-4bit-nibbleized, then sent as a sysex memory dump with the new values (cc parameter number of the clicked modulator, and the value of the "new" previous parameter changed)

So the path is cumbersome but it should work, it's not like a realtime midi or sysex parameter change but it's better than nothing. I just have to use a user preset as a buffer program (Program100) it's like a proxy.

I already decoded every parameter byte position for the 40 algos and I'm looking to translate those values to the GUI hence my question.

I need a double translation of a value:
the value received let's say 4 (out of 10)
it must be displayed as "-16 dB"
but when I click the slider the cc sent must be 128/10 x 4 = 51 (for a 128 scale)

Once I'll get the method right, it will be easy (but time consuming) to setup every value from tables and the value to CC method calculation.

I wonder if I should skip the CC control step and just send "almost" realtime sysex dumps of 300 bytes with a regular time interval of something acceptable like 0.5 sec. That would skip the step for value calculation to a 127 scale for the cc message.


Wish me good luck :)

**Reply by dobo365** [reply]
I didn't take a lot of time to understand deeply your need Damien but here are a few hints that I used for example in my Prologue panel.
On the PL, not all parameters have CC but all are available in sysex. I color coded those so the user can see which one have sysex and which ones have CC. At several places in my panel I have added small Send buttons that are directly sending the complete sysex to the synth. So not real time action but sending a sysex is fast so the user can directly hear the effect of the change

About your Midi values and modulator values:
Are you ware of the properties Expression to evaluate when calculating the modulator value from the midi message value and its sibling property for the opposite? Maybe you can use that?
Another way would be to link all your parameters to a common method that would send messages then on reception to use MidiMessageReceived to handle the value. Like a coded bidirectional behavior...

**Reply by damiensellier** [reply]
Thanks @dobo365 
Yes I see what you mean. This "send" button aside the slider  could be a solution as well. It's a nice idea.
I'll use it worst case scenario if the delay is too long between messages.

I went with a lookup table like this:

```
PreFxParamValue01 = SysexDecodedMessage:getByte(75)

Thrs={[0]=232,[1]=233,[2]=234,[3]=235,[4]=236,[5]=237,[6]=238,[7]=239,[8]=240,[9]=241,[10]=242,[11]=243,[12]=244,[13]=245,[14]=246,[15]=247,[16]=248,[17]=249,[18]=250,[19]=251,[20]=252,[21]=253,[22]=254,[23]=255,[24]=0,[25]=1,[26]=2,[27]=3,[28]=4,[29]=5,[30]=6,[31]=7,[32]=8,[33]=9,[34]=10,[35]=11,[36]=12}
	for k,v in pairs(Thrs) do
		if PreFxParamValue01 == v then
			PreFxParam01:getComponent():setValue(k, true)
		end
	end
```

And in my slider content I have 

`-24 dB
-23 dB
-22 dB
-21 dB
-20 dB
-19 dB
-18 dB
-17 dB
-16 dB
-15 dB
-14 dB
-13 dB
-12 dB
-11 dB
-10 dB
-09 dB
-08 dB
-07 dB
-06 dB
-05 dB
-04 dB
-03 dB
-02 dB
-01 dB
00 dB
+01 dB
+02 dB
+03 dB
+04 dB
+05 dB
+06 dB
+07 dB
+08 dB
+09 dB
+10 dB
+11 dB
+12 dB`


so the value to display is the key index corresponding to the value received.

for 232, the corresponding index of table is 0 so the mappedvalue is the first of the list : "-24 dB". It works great so far.
This one was tricky because of the negative values on 255 and under. But 99% are regular positive values so I don't need a table everytime, and I can directly use the slider content and its position from 0.

When the user will click on the slider, the CC value will be calculated on the fly and I'll store the calculated corresponding value as a "modulatorCustomIndex" for exemple. (it can be via a reverse table)

`PreFxParam01:setProperty("modulatorCustomIndex",SysexValueForDump, true)`

This way I'll get my 3 values correct:
The one received and translated, 
the one to send via CC on 0-127 scale 
the 3rd one to make the full memory sysex message.


The formula to send the CC value, will be an easy common value for all parameters.
I'll use expression to have the right value for each position.

The sysex message will be more complex and requires a magic formula that will translate all the edited parameters back to the machine.

It should be alright ;)
